src/azure-cli-core/azure/cli/core/util.py

```diff
@@ -74,78 +73,165 @@ def handle_exception(ex):  # pylint: disable=too-many-return-statements, too-man
-                    error_msg += ('\n' + detail)
-            except Exception:  # pylint: disable=broad-except
-                pass
-            az_error = AzCLIError(AzCLIErrorType.ValidationError, error_msg)
-            exit_code = ex.args[1] if len(ex.args) >= 2 else 1
-
-        # TODO: Fine-grained analysis
+        # TODO: AzureError is too general, need confirmation
+        elif isinstance(ex, AzureError):
+            if extract_common_error_message(ex):
+                error_msg = extract_common_error_message(ex)
+            AzCLIErrorType = get_error_type_by_azure_error(ex)
+            az_error = AzCLIErrorType(error_msg)
+
+        # TODO: AzureException is too general, need confirmation
         elif isinstance(ex, AzureException):
-            az_error = AzCLIError(AzCLIErrorType.ServiceError, error_msg)
-            exit_code = ex.args[1] if len(ex.args) >= 2 else 1
+            if is_azure_connection_error(error_msg):
+                az_error = azclierror.AzureConnectionError(error_msg)
+            else:
+                az_error = azclierror.ClientRequestError(error_msg)
 
-        # TODO: Fine-grained analysis
-        elif isinstance(ex, (ClientRequestError, SSLError)):
-            az_error = AzCLIError(AzCLIErrorType.ClientError, error_msg)
-            if 'SSLError' in error_msg:
-                az_error.set_recommendation(SSLERROR_TEMPLATE)
+        elif isinstance(ex, ClientRequestError):
+            if is_azure_connection_error(error_msg):
+                az_error = azclierror.AzureConnectionError(error_msg)
+            else:
+                az_error = azclierror.ClientRequestError(error_msg)
 
-        # TODO: Fine-grained analysis
         elif isinstance(ex, HttpOperationError):
-            try:
-                response = json.loads(ex.response.text)
-                if isinstance(response, str):
-                    error = response
-                else:
-                    error = response['error']
-
-                # ARM should use ODATA v4. So should try this first.
-                # http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091
-                if isinstance(error, dict):
-                    code = "{} - ".format(error.get('code', 'Unknown Code'))
-                    message = error.get('message', ex)
-                    error_msg = "code: {}, {}".format(code, message)
-                else:
-                    error_msg = error
-
-            except (ValueError, KeyError):
-                pass
-
-            az_error = AzCLIError(AzCLIErrorType.ServiceError, error_msg)
+            if extract_http_operation_error_message(ex):
+                error_msg = extract_http_operation_error_message(ex)
+            # by HttpOperationError definition, the message should be with the
+            # format: '({code}) {message}' if there is a status code
+            status_code = ex.__str__().split(')')[0].split('(')[-1]
```

Comment:
let's be defensive on this. :)

Reply:
If there is a different format, the statement will just return the ex.__str__(). It will not break the logic here, while i have to admit maybe there is a better way.